RFD.md

```diff
@@ -0,0 +1,217 @@
+* `StartCommand` - starts a command on the server and returns a unique `command_id` value used to manage the command in subsequent API calls.
+* `StopCommand` - stops a given command (identified by a `command_id`).
+* `CommandStatus` - returns current status information for a given command (identified by a `command_id`).
+* `CommandOutput` - a streaming API for receiving console output from a given command (identified by a `command_id`). If requested, the stream will continue until the command has finished (tail mode).
+
+The detailed GRPC definition for the proposed API can be found within the [remote_exec.proto](remote_exec/remote_exec.proto) file.
+
+### Library design
+
+Most of concerns around containerization will be isolated within the `container_exec` library, allowing us to have a very simple GRPC server around the library. This should make it easier to test the containerization code and contain most of the complexity (dealing with Linux APIs, process management, logs management, concurrency concerns, etc) the library.
+
+In a production scenario, this approach would make it possible to potentially reuse the library in multiple different systems (see ContainerD core services or the Moby project and its usage within Docker).
+
+#### Scope limits and design trade-offs
+
+We're going to apply the following design decisions while building the library:
+
+* Finished process state and logs will remain on the server and there will be no explicit way to delete those. It opens the server up to a number of DoS attacks and in a production scenario some kind of cleanup logic would need to be implemented (docker-like explicit delete command, time-based expiration, etc).
+
+* To avoid storing all output in memory while allowing us to stream command output from the beginning, we follow the approach used by Docker: log into a file per container and stream from the file as needed. The file is never deleted (see above for clarifications).
+
+* All processes started by the server share the same process group, so that we could kill them all as a group when killing the command.
+
+### CLI
+
+The client binary will be used as the primary mechanism for accessing the service during the exercise. In an actual production environment, we could potentially build different clients (web ui, cli, etc) all relying on the same fundamental GRPC API.
+
+The CLI will have the following commands available to the user:
+
+* `server-status` - shows remote server status
+* `run` - runs a remote command, waits for it to finish and streams it logs
+* `start` - runs a remote command asynchronously and returns its command id
+* `status` - shows the status of a given remote command
+* `kill` - stops a remote command
+* `logs` - shows remote command's console output (use `-tail` to follow the log)
```

Comment:
Will you be differentiating stdout and stderr or combining them?

Reply:
I was planning to combine both into a single stream. Clarified it in the API section now.